File: Source/Utility/ErrorHandler.cs
using System;

public static class ErrorHandler
{
    public static bool ErrorMessagesOn = true;
    public static Action Wrap(Action cb, Action<Exception>? errorCb = null)
    {
        return () =>
        {
            try
            {
                cb();
            }
            catch (Exception e)
            {
                if (ErrorMessagesOn) Logger.Warning("Error caught: " + e.Message + "\n" + e.StackTrace);
                errorCb?.Invoke(e);
            }
        };
    }
}

File: src/Utility/ErrorHandler.ts
import { Logger } from 'src/Events/Logger/Logger'
import { ErrorMessagesOn } from './ErrorMessagesOn'

export class ErrorHandler {
    private static callbackMap = new Map<() => unknown, () => unknown>()

    public static Wrap<T>(cb: () => T, errorCb?: (e: unknown) => void): () => unknown {
        if (ErrorHandler.callbackMap.has(cb)) {
            return ErrorHandler.callbackMap.get(cb)!
        }

        const wrapped = () => {
            const [ok, result] = pcall(cb)

            if (ok) {
                return result
            } else {
                if (ErrorMessagesOn.active) {
                    Logger.Warning(`Error caught: ${result}`)
                }
                errorCb?.(result)
            }
        }

        ErrorHandler.callbackMap.set(cb, wrapped)
        return wrapped
    }
}
